src/controller.cpp

```diff
@@ -0,0 +1,1230 @@
+       (-2 * c_lat_ * c_lat_ + 2) * controller_state_.filtered_error_deriv_lat.at(1));
+
+  controller_state_.error_deriv_ang.at(2) = controller_state_.error_deriv_ang.at(1);
+  controller_state_.error_deriv_ang.at(1) = controller_state_.error_deriv_ang.at(0);
+  controller_state_.error_deriv_ang.at(0) =
+      (controller_state_.error_ang.at(0) - controller_state_.error_ang.at(1)) / dt.toSec();
+  controller_state_.filtered_error_deriv_ang.at(2) = controller_state_.filtered_error_deriv_ang.at(1);
+  controller_state_.filtered_error_deriv_ang.at(1) = controller_state_.filtered_error_deriv_ang.at(0);
+  controller_state_.filtered_error_deriv_ang.at(0) =
+      (1 / (1 + c_ang_ * c_ang_ + M_SQRT2 * c_ang_)) *
+      (controller_state_.error_deriv_ang.at(2) + 2 * controller_state_.error_deriv_ang.at(1) +
+       controller_state_.error_deriv_ang.at(0) -
+       (c_ang_ * c_ang_ - M_SQRT2 * c_ang_ + 1) * controller_state_.filtered_error_deriv_ang.at(2) -
+       (-2 * c_ang_ * c_ang_ + 2) * controller_state_.filtered_error_deriv_ang.at(1));
+
+  // calculate the control effort
+  proportional_lat_ = Kp_lat_ * controller_state_.filtered_error_lat.at(0);
+  integral_lat_ = Ki_lat_ * controller_state_.error_integral_lat;
+  derivative_lat_ = Kd_lat_ * controller_state_.filtered_error_deriv_lat.at(0);
+
+  proportional_ang_ = Kp_ang_ * controller_state_.filtered_error_ang.at(0);
+  integral_ang_ = Ki_ang_ * controller_state_.error_integral_ang;
+  derivative_ang_ = Kd_ang_ * controller_state_.filtered_error_deriv_ang.at(0);
+
+
+  /***** Compute forward velocity *****/
+  // Apply acceleration limits and end velocity
+  double current_target_acc;
+  double acc;
+  double t_end_phase_current;
+  double d_end_phase;
+
+  // Compute time to reach end velocity from current velocity
+  // Compute estimate overall distance during end_phase
+  // The estimates are done a bit conservative to account that robot will take longer
+  // to de-accelerate and thus avoid abrupt velocity changes at the end of the trajectory
+  // The sample time plays an important role on how good these estimates are.
+  // Thus We add a distance to the end phase distance estimation depending on the sample time
+  if (current_x_vel > target_end_x_vel)
+  {
+    t_end_phase_current = (target_end_x_vel - current_x_vel) / (-target_x_decc_);
+    d_end_phase = current_x_vel * t_end_phase_current
+                - 0.5 * (target_x_decc_) * t_end_phase_current * t_end_phase_current
+                + target_x_vel * 2.0 * dt.toSec();
+  }
+  else
+  {
+    t_end_phase_current = (target_end_x_vel - current_x_vel) / (target_x_acc_);
+    d_end_phase = current_x_vel * t_end_phase_current
+                + 0.5 * (target_x_acc_) * t_end_phase_current * t_end_phase_current
+                + target_x_vel * 2.0 * dt.toSec();
+  }
+  ROS_DEBUG("t_end_phase_current: %f", t_end_phase_current);
+  ROS_DEBUG("d_end_phase: %f", d_end_phase);
+  ROS_DEBUG("distance_to_goal: %f", distance_to_goal_);
```

Comment:
I'm guessing right now that it is the distance to the last pose in the plan appended with a pose 1 carrot distance away from the last pose in the original plan.

Reply:
This is the distance to goal along the path, we should indeed use a different name.

Reply:
There is a mismatch of the behavior of this variable in the two control modes (trackin_base_link), which should be solved. Issue will be created.

Reply:
Problem is now that it is the distance along the path to the last pose in the global plan, which may be appended with another pose depending on the control strategy.

As discussed, candidate improvements are:
 - Do not append the global plan, but rather extrapolate the end at runtime (if necessary based on the specific control strategy at the time), this would solve the bug referred to in this conversation.
 - Maybe we could refactor the calculations in the two different control strategies to reduce code duplication.

Reply:
See #83